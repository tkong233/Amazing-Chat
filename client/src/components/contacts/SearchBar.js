import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { makeStyles } from '@material-ui/core/styles';
import SearchIcon from '@material-ui/icons/Search';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Divider from '@material-ui/core/Divider';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';

import ContactCard  from './ContactCard';
import SuggestionCard from './SuggestionCard';
import { getAllUsers } from '../../actions/contactActions';

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    backgroundColor: theme.palette.background.paper,
  },
  inline: {
    display: 'inline',
  },
  button: {
    maxWidth: '5px'
  }
}));

const SearchBar = (props) => {
  const classes = useStyles();

  useEffect(() => {
    props.getAllUsers();
  }, [])

  const [value, setValue] = useState('');
  const [result, setResult] = useState([]);

  const onChange = (e) => {
    const newValue = e.target.value;
    setValue(newValue);
    setResult(newValue == '' ? [] : props.users.filter(
      (user) => {
        return user.name.toLowerCase().indexOf(newValue.toLowerCase()) !== -1;
      }
    ));
  }

  const findInContact = (user) => {
    return props.contacts.filter(c => c.name === user.name)
  }

  return (
    <div data-test = "SearchBarComponent">
      {/* Search Input */}
    <div className="input-field">
      <input
      onChange={onChange}
      value={value}
      id="search"
      type="text"
    />
      <label htmlFor="search">Search</label>
    </div>

    {/* Search Result */}
    {value === '' ? null : (
      <List className={classes.root} alignItems="flex-start">
        <ListItem>
          <ListItemIcon>
            <SearchIcon />
          </ListItemIcon>
          <ListItemText primary="Search Result" />
        </ListItem>
        <Divider/>
          {result.map(r => {
            const contact = findInContact(r);
            return (
            contact.length > 0 ?
              <ContactCard
                name={contact[0].name}
                profilePicture={contact[0].profile_picture}
                lastInteractTime={contact[0].lastInteractTime}
                userEmail={props.user.email}
                contactEmail={contact[0].email}
                pairId={contact[0].pairId}
              />
              :
              <SuggestionCard
                name={r.name}
                profilePicture={r.profile_picture}
                userEmail={props.user.email}
                suggestionEmail={r.email}
              />
          )})}
        </List>
    )}
    
    </div>
  );
}

const mapStateToProps = state => ({
  user: state.auth.user,
  users: state.contact.users,
  contacts: state.contact.contacts,
})

export default connect(
  mapStateToProps,
  { getAllUsers }
)(SearchBar)