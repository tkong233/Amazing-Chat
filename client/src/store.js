import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from 'redux-persist'
import rootReducer from "./reducers";
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web

const initialState = {};

const persistConfig = {
  key: 'root',
  storage,
}

const middleware = [thunk];

const persistedReducer = persistReducer(persistConfig, rootReducer, initialState);

const store = createStore(

  persistedReducer,
  compose(
    applyMiddleware(...middleware),
    (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()) ||
      compose
  )
);

const persistor = persistStore(store);

export {
  store,
  persistor,
  middleware
};



