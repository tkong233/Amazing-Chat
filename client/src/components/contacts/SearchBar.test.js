import React from 'react';
import SearchBar from './SearchBar';
import {shallow} from 'enzyme';
import checkPropTypes from 'check-prop-types';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16'; 
import { applyMiddleware, createStore } from 'redux';
import {middleware} from '../../store';
import rootReducer from '../../reducers/index'

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const setUp = (initialState={}) =>{
    const auth = {
        user:{
            name: "testuser"
        }
    }
    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    const store = createStoreWithMiddleware(rootReducer, initialState);
    const wrapper = shallow(<SearchBar store={store} auth={auth}/>).dive();
    // console.log(wrapper.debug());
    return wrapper;
}

describe('SearchBar Component', ()=>{
    describe('Checking PropTypes', ()=>{
        it('Should not throw a warning', ()=>{
            const expectedProps = {
                contact: {
                    users:{},
                    contacts:{}
                },
                auth: {
                    user:{}
                },
            }
            // eslint-disable-next-line react/forbid-foreign-prop-types
            const propsErr = checkPropTypes(SearchBar.propTypes, expectedProps, 'props', SearchBar.name);
            expect(propsErr).toBeUndefined();
        })
    });
    describe('Renders', ()=>{
        let wrapper;
        beforeEach(()=>{
            const initialState = {
                auth:{user: {
                    name: 'testuser',
                    email: 'testuser@test.com',
                    profile_picture: 'testpicture'
                }}
            };
            wrapper = setUp(initialState);
        });

        it('Should render without errors', ()=>{
            const container = wrapper.find(`[data-test='SearchBarComponent']`);
            expect(container.length).toBe(1);
        });

        it('search a person', ()=>{
            const value = 'testUser'
            wrapper.find('input').at(0).simulate('change', {
                preventDefault: () => {},
                target: {
                    value
                }
            });
            console.log(wrapper);
            expect(wrapper.toBeUndefined);
        });

        it('Snapshot testing', ()=>{
            expect(wrapper).toMatchSnapshot();
        });


    });
});