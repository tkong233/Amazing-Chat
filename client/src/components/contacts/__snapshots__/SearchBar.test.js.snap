// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component Renders Snapshot testing 1`] = `ShallowWrapper {}`;
